modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/Sink.scala

```diff
@@ -148,67 +176,48 @@ object Sink {
-          Logger[F]
-            .error(exception)(
-              s"Writing ${records.size} records to ${config.streamName} errored (${retryDetails.retriesSoFar} retries from cats-retry)"
-            )
-      )
-
-    for {
-      _ <- retryStatus match {
-             case RetryStatus.NoRetriesYet =>
-               Logger[F].debug(s"Writing ${records.size} records to ${config.streamName}")
-             case retry =>
-               Logger[F].debug(
-                 s"Waiting for ${retry.cumulativeDelay} and retrying to write ${records.size} records to ${config.streamName}"
-               ) *>
-                 Timer[F].sleep(retry.cumulativeDelay)
-           }
-      putRecordsResult <- withRetry
-      failuresRetried <- if (putRecordsResult.getFailedRecordCount != 0) {
-                           val failurePairs = records.zip(putRecordsResult.getRecords.asScala).filter(_._2.getErrorMessage != null)
-                           val (failedRecords, failedResults) = failurePairs.unzip
-                           val logging = logErrorsSummary(getErrorsSummary(failedResults))
-                           val maybeRetrying =
-                             for {
-                               nextRetry <- retryPolicy.decideNextRetry(retryStatus)
-                               maybeRetried <- nextRetry match {
-                                                 case PolicyDecision.DelayAndRetry(delay) =>
-                                                   writeToKinesis(
-                                                     blocker,
-                                                     config,
-                                                     kinesis,
-                                                     failedRecords,
-                                                     retryPolicy,
-                                                     retryStatus.addRetry(delay)
-                                                   )
-                                                 case PolicyDecision.GiveUp =>
-                                                   Sync[F].raiseError[Unit](
-                                                     new RuntimeException(
-                                                       s"Maximum number of retries reached for ${failedRecords.size} records"
-                                                     )
-                                                   )
-                                               }
-                             } yield maybeRetried
-                           logging *> maybeRetrying
-                         } else
-                           Sync[F].unit
-    } yield failuresRetried
-  }
+    retryPolicy: RetryPolicy[F]
+  ): F[List[PutRecordsRequestEntry]] =
+    Logger[F].debug(s"Writing ${records.size} records to ${config.streamName}") *>
+      blocker
+        .blockOn(Sync[F].delay(putRecords(kinesis, config.streamName, records)))
+        .retryingOnFailuresAndAllErrors(
```

Comment:
> in case of one failure (e.g. throughput exceeded), isn't the retry executed on the exact same

No, because if there is one failure, then the `wasSuccessful` lambda returns true, so it does not get retried.  In this function, we only retry under these circumstances:

- An exception.  E.g. a http 5xx, a tcp connection error, anything else unexpected
- Every single record yielded an InternalFailure

Both of those cases, I want the app to crash if we exceed `maxRetries`, so we retry them in this block, not in the `runOnce` block, which uses a different retry policy.

Reply:
I see now. I was assuming that all the records that were returned by `tryWriteToKinesis` had met throughput exceeded error and had to be retried by `runOnce`, but now I understand that if only just one record of a batch gets this error, the whole batch is retried by `runOnce`.

I think that this is not right, and that instead the logic should be: unless all the records are `ThroughputExceeded`, we retry, until `maxRetries`. It's fine if we retry `ThroughputExceeded` errors, and `InternalErrorFailure`s should be retried. This way only records that met `ThroughputExceeded` (once or several times) will get retried by `runOnce`..

Reply:
> now I understand that if only just one record of a batch gets this error, the whole batch is retried by runOnce.

This is not how I implemented it.  If only one record of a batch gets the ThroughputExceeded error, then `tryWriteToKinesis` returns a list comprising just that single entry.  So the whole batch is not retried; just the single failure.

Honestly I am 100% convinced the logic is correct.  I think my mistake here is I have not made it clear enough what I am doing.  The fact that I'm having difficulty explaining it is proof that it is not very clear.  Bear with me, I will try to write it in a way that uses the same logic but is much more clear about where and why the retries happen.